core/src/main/java/com/linecorp/armeria/common/stream/FuseableStreamMessage.java

```diff
@@ -0,0 +1,324 @@
+        private static final AtomicReferenceFieldUpdater<FuseableSubscriber, Subscription> upstreamUpdater =
+                AtomicReferenceFieldUpdater
+                        .newUpdater(FuseableSubscriber.class, Subscription.class, "upstream");
+
+        private final Subscriber<? super U> downstream;
+        private final List<MapperFunction<Object, Object>> functions;
+
+        @Nullable
+        private volatile Subscription upstream;
+
+        FuseableSubscriber(Subscriber<? super U> downstream, List<MapperFunction<Object, Object>> functions) {
+            requireNonNull(downstream, "downstream");
+            this.downstream = downstream;
+            this.functions = functions;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            if (!upstreamUpdater.compareAndSet(this, null, subscription)) {
+                subscription.cancel();
+                throw new IllegalStateException("Subscription already set!");
+            }
+            upstream = subscription;
+            downstream.onSubscribe(this);
+        }
+
+        @Override
+        public void onNext(Object item) {
+            requireNonNull(item, "item");
+
+            final Subscription upstream = this.upstream;
+            if (upstream != null) {
+
+                Object result = item;
+                try {
+                    for (MapperFunction<Object, Object> function : functions) {
+                        assert result != null;
+                        final Type type = function.type();
+                        if (type == Type.MAP) {
+                            result = function.apply(result);
+                        } else if (type == Type.FILTER) {
+                            final Object filtered = function.apply(result);
+                            if (filtered == null) {
+                                // The given item was filtered out. Should request the next item.
+                                StreamMessageUtil.closeOrAbort(result, null);
+                                result = null;
+                                break;
+                            }
+                        } else {
+                            // Should never reach here.
+                            throw new Error();
+                        }
+                    }
+                } catch (Throwable ex) {
+                    upstream.cancel();
+                    onError(ex);
+                    return;
+                }
+
+                if (result != null) {
+                    @SuppressWarnings("unchecked")
+                    final U cast = (U) result;
+                    downstream.onNext(cast);
+                } else {
+                    upstream.request(1);
+                }
+            }
+        }
+
+        @Override
+        public void onError(Throwable throwable) {
+            if (upstream != null) {
```

Comment:
How about just calling `downstream.onError(throwable);` so that we don't access the volatile field twice?